src/ProducerEventProcessor/include/ProducerEventProcessor/UploaderClientCaptureEventCollector.h

```diff
@@ -34,6 +34,10 @@ class UploaderClientCaptureEventCollector final
   // file, and then buffer the converted data.
   void AddEvent(orbit_grpc_protos::ClientCaptureEvent&& event) override;
 
+  // Do nothing when stop and wait are requested by external. This collector will stop automatically
+  // after receiving and processing the `CaptureFinishedEvent`.
+  void OnStopAndWaitRequestedByExternal() override {}
```

Comment:
`UploaderClientCaptureEventCollector` has a private `Stop` method. It does not need to wait for the external request to stop.

Reply:
As far as I understand it, `StopAndWait` exists in the `GrpcClientCaptureEventCollector` to stop the capture processing before the `CaptureFinishedEvent` arrives. This is needed to stop the capture right away if the service runs out of memory.

Have you considered doing the same in the `UploaderClientCaptureEventCollector`? Not sure if it makes sense though, but it looks to me as if we could also have a `StopAndWait` function that sets an internal `stop_requested_` boolean to `true` and calls the private `Stop` method. In the `AddEvent` method we will check whether `stop_requested_` is `true` and discard all events if so.

Does that make sense?

Reply:
Thanks for pointing this out!

To explain the existing code a little bit:

The `StopAndWait` is called in `FinalizeEventProcessing` right after calling `producer_event_processor_->ProcessEvent(...)`. And in  `producer_event_processor_->ProcessEvent(...)`, event is converted into `ClientCaptureEvent` and sent to the `client_capture_event_collector` by calling `client_capture_event_collector->AddEvent(...)`.

For `UploaderClientCaptureEventCollector`, we do not have `stop_requested_`, and we call an internal `Stop` method inside `client_capture_event_collector->AddEvent(...)` when it realises a `CaptureFinishedEvent` is added. Hence, the `UploaderClientCaptureEventCollector` stops before we call `StopAndWait`.

And the `GrpcClientCaptureEventCollector` is stopped by `StopAndWait` after `producer_event_processor_->ProcessEvent(...)` processes a `CaptureFinishedEvent`.

Reply:
I'm thinking about that change `UploaderClientCaptureEventCollector::Stop` to `StopAndWait` and call it at the same place as the `GrpcClientCaptureEventCollector` case. WDYT?